program-runtime/src/loaded_programs.rs

```diff
@@ -235,9 +235,33 @@ impl LoadedPrograms {
                 return LoadedProgramEntry::WasOccupied(existing.clone());
             }
         }
-        let new_entry = Arc::new(entry);
-        second_level.insert(index.unwrap_or(second_level.len()), new_entry.clone());
-        LoadedProgramEntry::WasVacant(new_entry)
+        second_level.insert(index.unwrap_or(second_level.len()), entry.clone());
+        LoadedProgramEntry::WasVacant(entry)
+    }
+
+    /// Insert or replace the current entry with a tombstone.
+    /// This behaves differently than `insert_entry()` in case there's currently a program at the
+    /// `deploytment_slot`. It'll replace the current entry with a tombstone, whereas `insert_entry()`
+    /// would retain and return the current entry.
+    pub fn set_tombstone(&mut self, key: Pubkey, deployment_slot: Slot) -> Arc<LoadedProgram> {
```

Comment:
It does not feel like the right abstraction, because all forms of un/re/deployment should use the same insert method, including tombstones.

However, the current insert method covers both: Normal loading for execution and program-management-instructions. So maybe that should be split into two insert methods. One for program-management-instructions which can only add entries and never overwrite. And one for normal execution which can unify duplicate loads.